drivers/charger/bq2407x.c

```diff
@@ -0,0 +1,203 @@
+		return;
+	}
+
+	/*
+	 * The status table is as follows [stat1,stat2]
+	 *   [1,1] CHARGER_STATUS_CHARGING (precharge)
+	 *   [1,0] CHARGER_STATUS_CHARGING (fast charge)
+	 *   [0,1] CHARGER_STATUS_FULL
+	 *   [0,0] CHARGER_STATUS_NOT_CHARGING
+	 */
+	stat1 = gpio_pin_get_dt(&config->stat1_pin);
+	stat2 = gpio_pin_get_dt(&config->stat2_pin);
+	if ((stat1 && stat2) || stat1) {
+		val->status = CHARGER_STATUS_CHARGING;
+	} else if (stat2) {
+		val->status = CHARGER_STATUS_FULL;
+	} else {
+		val->status = CHARGER_STATUS_NOT_CHARGING;
+	}
+}
+
+static void ti_bq2407x_get_prop_charge_type(const struct device *dev, union charger_propval *val)
+{
+	const struct ti_bq2407x_config *config = dev->config;
+
+	if (gpio_pin_get_dt(&config->ce_pin)) {
+		val->charge_type = CHARGER_CHARGE_TYPE_ADAPTIVE;
+	} else {
+		val->charge_type = CHARGER_CHARGE_TYPE_NONE;
+	}
+}
+
+static int ti_bq2407x_get_property(const struct device *dev, const charger_prop_t prop,
+				   union charger_propval *val)
+{
+	struct ti_bq2407x_data *data = dev->data;
+	int ret = -ENOTSUP;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	switch (prop) {
+	case CHARGER_PROP_ONLINE:
+		ti_bq2407x_get_prop_online(dev, val);
+		ret = 0;
+		break;
+
+	case CHARGER_PROP_STATUS:
+		ti_bq2407x_get_prop_status(dev, val);
+		ret = 0;
+		break;
+
+	case CHARGER_PROP_CHARGE_TYPE:
+		ti_bq2407x_get_prop_charge_type(dev, val);
+		ret = 0;
+		break;
+
+	default:
+		break;
+	}
+
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+static int ti_bq2407x_set_prop_charge_type(const struct device *dev,
+					   const union charger_propval *val)
+
+{
+	const struct ti_bq2407x_config *config = dev->config;
+
+	switch (val->charge_type) {
+	case CHARGER_CHARGE_TYPE_NONE:
+		/* Disable charging */
+		gpio_pin_set_dt(&config->ce_pin, 0);
+		break;
+
+	case CHARGER_CHARGE_TYPE_ADAPTIVE:
+		/* Enable charging */
+		gpio_pin_set_dt(&config->ce_pin, 1);
+		break;
+
+	default:
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int ti_bq2407x_set_property(const struct device *dev, const charger_prop_t prop,
+				    const union charger_propval *val)
+{
+	struct ti_bq2407x_data *data = dev->data;
+	int ret = -ENOTSUP;
+
+	k_spinlock_key_t key = k_spin_lock(&data->lock);
+
+	switch (prop) {
+	case CHARGER_PROP_CHARGE_TYPE:
+		ret = ti_bq2407x_set_prop_charge_type(dev, val);
+		break;
+
+	default:
+		break;
+	}
+
+	k_spin_unlock(&data->lock, key);
+	return ret;
+}
+
+static const struct charger_driver_api ti_bq2407x_api = {
+	.get_property = ti_bq2407x_get_property,
+	.set_property = ti_bq2407x_set_property,
+};
+
+static int ti_bq2407x_init(const struct device *dev)
+{
+	struct ti_bq2407x_config *config = (struct ti_bq2407x_config *)dev->config;
+
+	if ((gpio_pin_configure_dt(&config->stat1_pin, 0) < 0) ||
+	    (gpio_pin_configure_dt(&config->stat2_pin, 0) < 0) ||
+	    (gpio_pin_configure_dt(&config->pg_pin, 0) < 0)) {
+		return -ENODEV;
+	}
+
+#ifdef CONFIG_BQ2407X_INIT_ENABLE_CHARGING
```

Comment:
I feel like this is common of a feature that it could be a common charger dt property.

Would you mind writing a `charger.yaml` in `dts/bindings/charger/`?

Reply:
I will definitely add it :)

Reply:
Added